### DISCUSSION TITLE
Can atomic with barrier be used to ensure visibility between different workgroups?

### DISCUSSION THREAD

**Discussion by AntiAnimeGeneral** [gpuweb/gpuweb#4821] [Q&A] [answered]
I know storageBarrier  can't sync workgroups, But can it ensure visibility between different workgroups?
example
```
X=0;
Y=0;
--------------
workgroup1:
{
    X=777;
    storageBarrier(); // rel
    atomicStore(&Y,1);
}
workgroup2:
{
    while atomicLoad(&Y)==0{}
    storageBarrier(); // acq
    let _x=X;
}
```
As I know storageBarrier() has acq_rel semantics, but I'm not sure if the acq_rel semantics apply to the grid
Can I ensure that _x is equal to 777?

**Comment by alan-baker** [comment]
Unfortunately, barriers only have a memory scope of workgroup currently. That means the visibility effects do not extend past the workgroup. There is no way currently to increase that visibility for non-atomic memory operations.

Side note: the spin loop in workgroup 2 is non-portable and might well lead to a hang. There are insufficient forward progress guarantees to make it work.

**Reply by AntiAnimeGeneral** [reply]
If X is also an atomic variable, can the barrier ensure the ordered visibility of the two atomic variables?
```
X=0;
Y=0;
--------------
workgroup1:
{
    atomicStore(&X,777);
    storageBarrier(); // rel
    atomicStore(&Y,1);
}
workgroup2:
{
    while atomicLoad(&Y)==0{}
    storageBarrier(); // acq
    let _x=atomicLoad(&X);
}
```
Forgive me for still fantasizing about wgsl providing more synchronization guarantees

**Reply by Kangz** [reply]
I'm not sure if the above works, but you should be careful about it because Metal doesn't provide forward progress guarantees that would ensure that the code above works (even if you know that workgroup1 is launched before workgroup2). So even if things work on your machine, they might happen to deadlock on some other.

**Reply by AntiAnimeGeneral** [reply]
Okay, I don't necessarily have to use spin loops, it's just that it's easy to demonstrate, so I can get good news: barrier can ensure the visibility ord of atomic variables, right? (I really care about this)

**Reply by alan-baker** [reply]
If you make X atomic, you don't need the barrier or other atomic. Atomic operations on X are mutually ordered and storage atomic use a `QueueFamily` scope (basically equivalent to device).

**Reply by AntiAnimeGeneral** [reply]
Really? I am not very familiar with QueueFamily. To prevent misunderstandings, I need to make one final confirmation,
```
X=0;
Y=0;
--------------
workgroup1:
{
    atomicStore(&X,777);
    atomicStore(&Y,1);
}
workgroup2:
{
    let _y=atomicLoad(&Y);
    let _x=atomicLoad(&X);
}
```
Will the following situations occur:
_y==1 but   _x==0;
On the CPU, if I don't add mem_ord, this could happen

**Reply by alan-baker** [reply]
Ah yes, sorry I thought you were only concerned with reading X in different workgroups. Adding a storage barrier between the atomics in each workgroup would prevent reordering the operations.

**Reply by AntiAnimeGeneral** [reply]
so
```
X=0;
Y=0;
--------------
workgroup1:
{
    atomicStore(&X,777);
    storageBarrier();
    atomicStore(&Y,1);
}
workgroup2:
{
    let _y=atomicLoad(&Y);
    storageBarrier();
    let _x=atomicLoad(&X);
}
```
The following situations will never happen
```
_y==1 but _x==0;
```
If  _y==1,  According to acq_rel semantics, _x is 777

right?

**Reply by alan-baker** [reply, marked as answer]
Yes that should be the case.

**Reply by AntiAnimeGeneral** [reply]
Thank you, this can provide some cross workgroup consistency.
